GoogleForCreators/web-stories-wp
# Added skeleton thumbnails until first idle frame

## Context



## Summary

This adds non-interactive, background-only skeleton thumbnails while the editor is still loading. The real thumbnails are displayed as soon as the javascript engine either has an idle frame or 5 seconds have passed.

## Todo

There are four (potential) issues currently, that still need to be resolved.

* [x] `requestIdleCallback` is [not supported at all by Safari](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback#browser_compatibility). But as always [there's a polyfill](https://github.com/behnammodi/polyfill/blob/master/window.polyfill.js). How do we go about adding polyfills?
* [x] Is accessibility correct around this? Do we need to do anything special regarding accessibility for these temporary placeholders? They're marked with `role="presentation"` so I assume they're just completely ignored by screen readers, but should we somehow alert a screen reader, that the "real" thumbnails have loaded once they do?
* [x] Race condition and Percy: Do we risk that screenshots in Percy are now unstable because sometimes the "real" thumbnails will load, sometimes the skeleton ones will still be displayed, when the screenshot is taken? Is there something we can do to alleviate this? If we make sure to wait for the real thumbnails to be present on the page before taking screenshots, that has to happen in many places throughout the code then. Maybe we can implement it centrally somewhere - if it's even a problem?
* [x] Is the 5-second delay too small for slower laptops? Should we increase it to 10?

## User-facing changes

While loading, the thumbnails display only the background color, not all elements:
![skeleton-thumbnails](https://user-images.githubusercontent.com/637548/125874995-4478f403-a3dc-4341-98b3-586c46c3cfa3.gif)

## Testing Instructions

### QA

This PR can be tested by following these steps:

1. Load a story with many pages with both background colors and background media elements plus elements on all pages
2. Observe that while the story is loading, the thumbnails initially render only the background color or in case of background media the primary color of the background media and the thumbnails are also non-interactive in this state
3. Once the full editor and story has loaded, observe that the thumbnails are populated with full interaction and all elements displayed

Please **make sure to QA this on a slow(er) laptop** as well as high-end laptops. Validate that the thumbnails still show up eventually and don't clog things up.

### UAT




## Checklist




---

Fixes #8254
